pkg/logproto/extensions.go

```diff
@@ -16,39 +16,54 @@ import (
 var seps = []byte{'\xff'}
 
 // Hash returns hash of the labels according to Prometheus' Labels.Hash function.
-// `b` and `keysForLabels` are buffers that should be reused to avoid
-// allocations.
-func (id SeriesIdentifier) Hash(b []byte, keysForLabels []string) (uint64, []string) {
-	keysForLabels = keysForLabels[:0]
-	for k := range id.Labels {
-		keysForLabels = append(keysForLabels, k)
-	}
-	sort.Strings(keysForLabels)
+// `b` is a buffer that should be reused to avoid allocations.
+func (id SeriesIdentifier) Hash(b []byte) uint64 {
+	sort.Slice(id.Labels, func(l, r int) bool { return id.Labels[l].Key < id.Labels[r].Key })
 
 	// Use xxhash.Sum64(b) for fast path as it's faster.
 	b = b[:0]
-	for i, name := range keysForLabels {
-		value := id.Labels[name]
+	for i, pair := range id.Labels {
+		name := pair.Key
+		value := pair.Value
 		if len(b)+len(name)+len(value)+2 >= cap(b) {
 			// If labels entry is 1KB+ do not allocate whole entry.
 			h := xxhash.New()
 			_, _ = h.Write(b)
-			for _, name := range keysForLabels[i:] {
-				value := id.Labels[name]
+			for _, pair := range id.Labels[i:] {
+				name := pair.Key
+				value := pair.Value
 				_, _ = h.WriteString(name)
 				_, _ = h.Write(seps)
 				_, _ = h.WriteString(value)
 				_, _ = h.Write(seps)
 			}
-			return h.Sum64(), keysForLabels
+			return h.Sum64()
 		}
 
 		b = append(b, name...)
 		b = append(b, seps[0])
 		b = append(b, value...)
 		b = append(b, seps[0])
 	}
-	return xxhash.Sum64(b), keysForLabels
+	return xxhash.Sum64(b)
+}
+
+func (id SeriesIdentifier) Get(key string) string {
+	for _, entry := range id.Labels {
+		if entry.Key == key {
+			return entry.Value
+		}
+	}
+
+	return ""
+}
+
+func NewSeriesEntries(labels ...string) []SeriesIdentifier_LabelsEntry {
+	r := make([]SeriesIdentifier_LabelsEntry, 0, len(labels)/2)
```

Comment:
Maybe add a check if `len(labels)` is divisible by 2?

Reply:
It's only used in tests but I will add it.

Reply:
Added a panic following `labels.FromStrings`.

Reply:
If we're going to put a panic here, should we rename the function to something like `MustNewSeriesEntries`? This could help someone reading the code or if someone reaches for it outside of tests.

Reply:
Yes. Good point.